# Association Rules

In [15]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

### Preparing data set

In [16]:
customer_info = pd.read_csv("1st_clustering_results.csv", index_col=0)
customer_basket = pd.read_csv("customer_basket.csv", index_col=0)
customer_info_backup = pd.read_csv("customer_info_backup.csv", index_col=0)

In [17]:
outlier_ids = customer_info_backup.index.difference(customer_info.index)

In [18]:
print(len(outlier_ids))

3327


In [19]:
outliers_df = customer_info_backup.loc[outlier_ids].copy()
outliers_df['segmentation'] = 'Outliers'

In [20]:
customer_info_full = pd.concat([customer_info, outliers_df])

In [21]:
print(len(customer_info_full))


34060


In [22]:
customer_info_full[customer_info_full['segmentation'] != 'Outliers']

segmentation customer_name  customer_gender  \
customer_id                                                           
29930        Large Family Households           NaN              NaN   
6813          Promo-Oriented Seniors           NaN              NaN   
39451        Large Family Households           NaN              NaN   
21557              Tech-Savvy Buyers           NaN              NaN   
16415            Wide-Range Shoppers           NaN              NaN   
...                              ...           ...              ...   
9629                       Alfragide           NaN              NaN   
24185                      Alfragide           NaN              NaN   
10075                      Alfragide           NaN              NaN   
7703                       Alfragide           NaN              NaN   
459                        Alfragide           NaN              NaN   

             kids_home  teens_home  number_complaints  \
customer_id                                             
29930              NaN         NaN                NaN   
6813               NaN         NaN                NaN   
39451              NaN         NaN                NaN   
21557              NaN         NaN                NaN   
16415              NaN         NaN                NaN   
...                ...         ...                ...   
9629               NaN         NaN                NaN   
24185              NaN         NaN                NaN   
10075              NaN         NaN                NaN   
7703               NaN         NaN                NaN   
459                NaN         NaN                NaN   

             distinct_stores_visited  lifetime_spend_groceries  \
customer_id                                                      
29930                            NaN                       NaN   
6813                             NaN                       NaN   
39451                            NaN                       NaN   
21557                            NaN                       NaN   
16415                            NaN                       NaN   
...                              ...                       ...   
9629                             NaN                       NaN   
24185                            NaN                       NaN   
10075                            NaN                       NaN   
7703                             NaN                       NaN   
459                              NaN                       NaN   

             lifetime_spend_electronics  typical_hour  ...  \
customer_id                                            ...   
29930                               NaN           NaN  ...   
6813                                NaN           NaN  ...   
39451                               NaN           NaN  ...   
21557                               NaN           NaN  ...   
16415                               NaN           NaN  ...   
...                                 ...           ...  ...   
9629                                NaN           NaN  ...   
24185                               NaN           NaN  ...   
10075                               NaN           NaN  ...   
7703                                NaN           NaN  ...   
459                                 NaN           NaN  ...   

             lifetime_spend_hygiene  lifetime_spend_videogames  \
customer_id                                                      
29930                           NaN                        NaN   
6813                            NaN                        NaN   
39451                           NaN                        NaN   
21557                           NaN                        NaN   
16415                           NaN                        NaN   
...                             ...                        ...   
9629                            NaN                        NaN   
24185                           NaN                        NaN   
10075                           

Saving final clustering results including ouliers:

In [23]:
customer_info_full[['segmentation']].to_csv('clustering_results.csv')

In [24]:
basket_with_segment = customer_basket.merge(
    customer_info_full[['segmentation']], 
    left_on='customer_id', 
    right_index=True,
    how='left'
)

In [25]:
basket_with_segment

list_of_goods  customer_id  \
invoice_id                                                                   
5062209     ['megaman zero 3', 'energy bar', 'pokemon viol...         4925   
4272512     ['black beer', 'bramble', 'laptop', 'pokemon v...        19046   
7121052     ['soup', 'shallot', 'cake', 'fresh bread', 'cr...        10318   
5847748     ['soup', 'cooking oil', 'cereals', 'oil', 'dog...        27283   
6336114     ['eggplant', 'mint green tea', 'oil', 'bacon',...        16072   
...                                                       ...          ...   
5389466     ['frozen smoothie', 'avocado', 'dog food', 'ch...        26170   
10387390    ['cake', 'antioxydant juice', 'honey', 'toothp...        19691   
7103407     ['candy bars', 'spaghetti', 'oil', 'final fant...         1720   
4093515     ['shallot', 'cereals', 'oil', 'deodorant', 'me...        24989   
6611217     ['soup', 'napkins', 'pokemon violet', 'ratchet...          482   

                       segmentation  
invoice_id                           
5062209           Tech-Savvy Buyers  
4272512           Tech-Savvy Buyers  
7121052       Young Promo-Sensitive  
5847748      Promo-Oriented Seniors  
6336114         Wide-Range Shoppers  
...                             ...  
5389466                    Outliers  
10387390                Vegetarians  
7103407       Young Promo-Sensitive  
4093515      Promo-Oriented Seniors  
6611217     Large Family Households  

[100000 rows x 3 columns]

### Goods by segment

In [26]:
basket_with_segment.groupby('segmentation')['list_of_goods'].count()


segmentation
Alfragide                    274
High-Spending Shoppers     10234
Large Family Households    15888
Outliers                    9742
Promo-Oriented Seniors     14187
Tech-Savvy Buyers           9575
Vegetarians                12812
Wide-Range Shoppers        15150
Young Promo-Sensitive      12138
Name: list_of_goods, dtype: int64

In [27]:
segmenti = basket_with_segment['segmentation'].unique()
segment_rules = {}

for segment in segmenti:
    basket_seg = basket_with_segment[basket_with_segment['segmentation'] == segment]
    transactions = basket_seg["list_of_goods"].apply(eval).tolist()
    te = TransactionEncoder()
    te_ary = te.fit(transactions).transform(transactions)
    df_encoded = pd.DataFrame(te_ary, columns=te.columns_)

    frequent_itemsets = apriori(
        df_encoded, 
        min_support=0.01, 
        use_colnames=True, 
        max_len=2, 
        low_memory=True
    )
    
    rules = association_rules(
        frequent_itemsets, 
        metric="lift", 
        min_threshold=1.1
    ).sort_values("lift", ascending=False).reset_index(drop=True)
    
    segment_rules[segment] = rules


This code finds the strongest product associations for each customer segment, allowing you to see which products are commonly bought together within each group.

In [28]:
for segment in segment_rules:
    rules = segment_rules[segment].head(10)
    print(f"\nTop ruler for the segment {segment}")
    display(rules)



Top ruler for the segment Tech-Savvy Buyers


antecedents     consequents  antecedent support  consequent support  \
0    (white wine)         (cider)            0.188198            0.147990   
1         (cider)    (white wine)            0.147990            0.188198   
2  (dessert wine)    (white wine)            0.122507            0.188198   
3    (white wine)  (dessert wine)            0.188198            0.122507   
4         (cider)  (dessert wine)            0.147990            0.122507   
5  (dessert wine)         (cider)            0.122507            0.147990   
6       (bramble)  (dessert wine)            0.084282            0.122507   
7  (dessert wine)       (bramble)            0.122507            0.084282   
8     (champagne)    (white wine)            0.090966            0.188198   
9    (white wine)     (champagne)            0.188198            0.090966   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.092637    0.492231  3.326119               1.0  0.064786    1.677948   
1  0.092637    0.625970  3.326119               1.0  0.064786    2.170420   
2  0.070183    0.572890  3.044074               1.0  0.047127    1.900685   
3  0.070183    0.372919  3.044074               1.0  0.047127    1.399330   
4  0.053159    0.359210  2.932167               1.0  0.035030    1.369392   
5  0.053159    0.433930  2.932167               1.0  0.035030    1.505133   
6  0.028198    0.334572  2.731058               1.0  0.017873    1.318691   
7  0.028198    0.230179  2.731058               1.0  0.017873    1.189521   
8  0.046162    0.507463  2.696424               1.0  0.029042    1.648203   
9  0.046162    0.245283  2.696424               1.0  0.029042    1.204470   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.861478  0.380360   0.404034    0.559101  
1       0.820822  0.380360   0.539260    0.559101  
2       0.765240  0.291793   0.473874    0.472905  
3       0.827164  0.291793   0.285372    0.472905  
4       0.773412  0.244594   0.269749    0.396570  
5       0.750952  0.244594   0.335607    0.396570  
6       0.692180  0.157895   0.241672    0.282376  
7       0.722332  0.157895   0.159325    0.282376  
8       0.692096  0.198117   0.393279    0.376373  
9       0.774990  0.198117   0.169759    0.376373


Top ruler for the segment Young Promo-Sensitive


antecedents       consequents  antecedent support  consequent support  \
0  (barbecue sauce)       (deodorant)            0.044488            0.037898   
1       (deodorant)  (barbecue sauce)            0.037898            0.044488   
2         (cologne)  (barbecue sauce)            0.047207            0.044488   
3  (barbecue sauce)         (cologne)            0.044488            0.047207   
4            (beer)            (salt)            0.131817            0.069781   
5            (salt)            (beer)            0.069781            0.131817   
6      (white wine)            (beer)            0.130499            0.131817   
7            (beer)      (white wine)            0.131817            0.130499   
8           (cider)      (white wine)            0.097133            0.130499   
9      (white wine)           (cider)            0.130499            0.097133   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.011040    0.248148  6.547874               1.0  0.009354    1.279644   
1  0.011040    0.291304  6.547874               1.0  0.009354    1.348268   
2  0.011040    0.233857  5.256583               1.0  0.008940    1.247171   
3  0.011040    0.248148  5.256583               1.0  0.008940    1.267261   
4  0.042594    0.323125  4.630568               1.0  0.033395    1.374285   
5  0.042594    0.610390  4.630568               1.0  0.033395    2.228335   
6  0.070275    0.538510  4.085272               1.0  0.053073    1.881260   
7  0.070275    0.533125  4.085272               1.0  0.053073    1.862384   
8  0.051162    0.526718  4.036173               1.0  0.038486    1.837171   
9  0.051162    0.392045  4.036173               1.0  0.038486    1.485090   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.886728  0.154734   0.218532    0.269726  
1       0.880653  0.154734   0.258308    0.269726  
2       0.849883  0.136874   0.198185    0.241003  
3       0.847465  0.136874   0.210897    0.241003  
4       0.903086  0.267876   0.272349    0.466757  
5       0.842859  0.267876   0.551234    0.466757  
6       0.868565  0.365937   0.468441    0.535818  
7       0.869884  0.365937   0.463054    0.535818  
8       0.833169  0.289916   0.455685    0.459382  
9       0.865141  0.289916   0.326640    0.459382


Top ruler for the segment Promo-Oriented Seniors


antecedents     consequents  antecedent support  consequent support  \
0         (cider)    (white wine)            0.032283            0.036724   
1    (white wine)         (cider)            0.036724            0.032283   
2   (fresh bread)        (butter)            0.159019            0.048425   
3        (butter)   (fresh bread)            0.048425            0.159019   
4       (cereals)      (pet food)            0.093466            0.102629   
5      (pet food)       (cereals)            0.102629            0.093466   
6         (cream)      (pet food)            0.079791            0.102629   
7      (pet food)         (cream)            0.102629            0.079791   
8  (french fries)    (mayonnaise)            0.149292            0.060760   
9    (mayonnaise)  (french fries)            0.060760            0.149292   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.010362    0.320961  8.739865               1.0  0.009176    1.418587   
1  0.010362    0.282150  8.739865               1.0  0.009176    1.348076   
2  0.010362    0.065160  1.345588               1.0  0.002661    1.017901   
3  0.010362    0.213974  1.345588               1.0  0.002661    1.069915   
4  0.012406    0.132730  1.293297               1.0  0.002813    1.034708   
5  0.012406    0.120879  1.293297               1.0  0.002813    1.031183   
6  0.010291    0.128975  1.256712               1.0  0.002102    1.030247   
7  0.010291    0.100275  1.256712               1.0  0.002102    1.022766   
8  0.011278    0.075543  1.243304               1.0  0.002207    1.015991   
9  0.011278    0.185615  1.243304               1.0  0.002207    1.044602   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.915125  0.176683   0.295073    0.301555  
1       0.919344  0.176683   0.258202    0.301555  
2       0.305394  0.052575   0.017587    0.139567  
3       0.269900  0.052575   0.065346    0.139567  
4       0.250164  0.067536   0.033543    0.126805  
5       0.252719  0.067536   0.030240    0.126805  
6       0.221985  0.059787   0.029359    0.114625  
7       0.227634  0.059787   0.022259    0.114625  
8       0.230034  0.056738   0.015739    0.130579  
9       0.208351  0.056738   0.042698    0.130579


Top ruler for the segment Wide-Range Shoppers


antecedents     consequents  antecedent support  consequent support  \
0    (white wine)         (cider)            0.058152            0.042904   
1         (cider)    (white wine)            0.042904            0.058152   
2         (cider)  (dessert wine)            0.042904            0.038548   
3  (dessert wine)         (cider)            0.038548            0.042904   
4    (white wine)          (beer)            0.058152            0.033861   
5          (beer)    (white wine)            0.033861            0.058152   
6  (dessert wine)    (white wine)            0.038548            0.058152   
7    (white wine)  (dessert wine)            0.058152            0.038548   
8       (bramble)    (white wine)            0.032013            0.058152   
9    (white wine)       (bramble)            0.058152            0.032013   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.024092    0.414302  9.656422               1.0  0.021597    1.634111   
1  0.024092    0.561538  9.656422               1.0  0.021597    2.148075   
2  0.014257    0.332308  8.620653               1.0  0.012604    1.439963   
3  0.014257    0.369863  8.620653               1.0  0.012604    1.518869   
4  0.016370    0.281498  8.313254               1.0  0.014401    1.344657   
5  0.016370    0.483431  8.313254               1.0  0.014401    1.823276   
6  0.018548    0.481164  8.274280               1.0  0.016306    1.815311   
7  0.018548    0.318956  8.274280               1.0  0.016306    1.411732   
8  0.013597    0.424742  7.304024               1.0  0.011736    1.637263   
9  0.013597    0.233825  7.304024               1.0  0.011736    1.263402   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.951790  0.313036   0.388046    0.487920  
1       0.936627  0.313036   0.534467    0.487920  
2       0.923627  0.212181   0.305538    0.351085  
3       0.919442  0.212181   0.341615    0.351085  
4       0.934025  0.216405   0.256316    0.382465  
5       0.910542  0.216405   0.451537    0.382465  
6       0.914391  0.237331   0.449130    0.400060  
7       0.933424  0.237331   0.291650    0.400060  
8       0.891633  0.177586   0.389225    0.329284  
9       0.916378  0.177586   0.208486    0.329284


Top ruler for the segment Vegetarians


antecedents      consequents  antecedent support  consequent support  \
0      (flax seed)          (salad)            0.133703            0.076803   
1          (salad)      (flax seed)            0.076803            0.133703   
2        (shallot)     (body spray)            0.136513            0.070012   
3     (body spray)        (shallot)            0.070012            0.136513   
4     (shower gel)        (burgers)            0.116297            0.070091   
5        (burgers)     (shower gel)            0.070091            0.116297   
6        (cologne)   (green grapes)            0.101780            0.089838   
7   (green grapes)        (cologne)            0.089838            0.101780   
8  (metroid prime)         (melons)            0.037153            0.258508   
9         (melons)  (metroid prime)            0.258508            0.037153   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.013503    0.100992  1.314954               1.0  0.003234    1.026907   
1  0.013503    0.175813  1.314954               1.0  0.003234    1.051093   
2  0.012566    0.092053  1.314803               1.0  0.003009    1.024275   
3  0.012566    0.179487  1.314803               1.0  0.003009    1.052375   
4  0.010615    0.091275  1.302247               1.0  0.002464    1.023312   
5  0.010615    0.151448  1.302247               1.0  0.002464    1.041424   
6  0.011786    0.115798  1.288965               1.0  0.002642    1.029360   
7  0.011786    0.131190  1.288965               1.0  0.002642    1.033852   
8  0.012176    0.327731  1.267781               1.0  0.002572    1.102970   
9  0.012176    0.047101  1.267781               1.0  0.002572    1.010441   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.276484  0.068542   0.026202    0.138403  
1       0.259443  0.068542   0.048609    0.138403  
2       0.277282  0.064789   0.023699    0.135770  
3       0.257455  0.064789   0.049769    0.135770  
4       0.262641  0.060391   0.022781    0.121361  
5       0.249590  0.060391   0.039776    0.121361  
6       0.249586  0.065538   0.028522    0.123494  
7       0.246312  0.065538   0.032743    0.123494  
8       0.219370  0.042952   0.093357    0.187416  
9       0.284858  0.042952   0.010333    0.187416


Top ruler for the segment Outliers


antecedents        consequents  antecedent support  \
0         (pancakes)     (energy drink)            0.044960   
1     (energy drink)         (pancakes)            0.052043   
2        (minecraft)  (ratchet & clank)            0.051427   
3  (ratchet & clank)        (minecraft)            0.066003   
4      (babies food)  (ratchet & clank)            0.100082   
5  (ratchet & clank)      (babies food)            0.066003   
6        (minecraft)      (babies food)            0.051427   
7      (babies food)        (minecraft)            0.100082   
8         (tomatoes)          (carrots)            0.077807   
9          (carrots)         (tomatoes)            0.059433   

   consequent support   support  confidence      lift  representativity  \
0            0.052043  0.016732    0.372146  7.150784               1.0   
1            0.044960  0.016732    0.321499  7.150784               1.0   
2            0.066003  0.020735    0.403194  6.108728               1.0   
3            0.051427  0.020735    0.314152  6.108728               1.0   
4            0.066003  0.040135    0.401026  6.075881               1.0   
5            0.100082  0.040135    0.608087  6.075881               1.0   
6            0.100082  0.028126    0.546906  5.464574               1.0   
7            0.051427  0.028126    0.281026  5.464574               1.0   
8            0.059433  0.023712    0.304749  5.127579               1.0   
9            0.077807  0.023712    0.398964  5.127579               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.014392    1.509837       0.900648  0.208440   0.337677    0.346823  
1  0.014392    1.407574       0.907378  0.208440   0.289558    0.346823  
2  0.017341    1.564992       0.881640  0.214437   0.361019    0.358673  
3  0.017341    1.383067       0.895399  0.214437   0.276969    0.358673  
4  0.033530    1.559327       0.928323  0.318663   0.358698    0.504556  
5  0.033530    2.296219       0.894451  0.318663   0.564501    0.504556  
6  0.022979    1.986162       0.861297  0.227953   0.496516    0.413966  
7  0.022979    1.319342       0.907864  0.227953   0.242047    0.413966  
8  0.019087    1.352845       0.872894  0.208861   0.260817    0.351857  
9  0.019087    1.534338       0.855842  0.208861   0.348253    0.351857


Top ruler for the segment Large Family Households


antecedents        consequents  antecedent support  \
0             (milk)            (cream)            0.153575   
1            (cream)             (milk)            0.052933   
2          (burgers)  (sparkling water)            0.224635   
3  (sparkling water)          (burgers)            0.036883   
4      (cooking oil)   (metroid fusion)            0.357880   
5   (metroid fusion)      (cooking oil)            0.023225   
6         (escalope)        (olive oil)            0.200969   
7        (olive oil)         (escalope)            0.063633   
8             (milk)   (pokemon violet)            0.153575   
9   (pokemon violet)             (milk)            0.124182   

   consequent support   support  confidence      lift  representativity  \
0            0.052933  0.010008    0.065164  1.231064               1.0   
1            0.153575  0.010008    0.189061  1.231064               1.0   
2            0.036883  0.010133    0.045111  1.223069               1.0   
3            0.224635  0.010133    0.274744  1.223069               1.0   
4            0.023225  0.010070    0.028139  1.211591               1.0   
5            0.357880  0.010070    0.433604  1.211591               1.0   
6            0.063633  0.015483    0.077044  1.210749               1.0   
7            0.200969  0.015483    0.243323  1.210749               1.0   
8            0.124182  0.022910    0.149180  1.201306               1.0   
9            0.153575  0.022910    0.184491  1.201306               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.001878    1.013083       0.221750  0.050929   0.012915    0.127112  
1  0.001878    1.043759       0.198185  0.050929   0.041924    0.127112  
2  0.001848    1.008616       0.235224  0.040310   0.008543    0.159927  
3  0.001848    1.069092       0.189369  0.040310   0.064626    0.159927  
4  0.001759    1.005057       0.271972  0.027142   0.005031    0.230872  
5  0.001759    1.133695       0.178791  0.027142   0.117928    0.230872  
6  0.002695    1.014530       0.217846  0.062153   0.014322    0.160183  
7  0.002695    1.055974       0.185894  0.062153   0.053007    0.160183  
8  0.003839    1.029382       0.197977  0.089899   0.028543    0.166835  
9  0.003839    1.037910       0.191333  0.089899   0.036525    0.166835


Top ruler for the segment High-Spending Shoppers


antecedents     consequents  antecedent support  consequent support  \
0         (cider)    (white wine)            0.036545            0.043189   
1    (white wine)         (cider)            0.043189            0.036545   
2    (white wine)  (dessert wine)            0.043189            0.033027   
3  (dessert wine)    (white wine)            0.033027            0.043189   
4     (meatballs)   (blueberries)            0.057846            0.132011   
5   (blueberries)     (meatballs)            0.132011            0.057846   
6      (hot dogs)        (laptop)            0.044069            0.200410   
7        (laptop)      (hot dogs)            0.200410            0.044069   
8   (french wine)        (shrimp)            0.104749            0.074849   
9        (shrimp)   (french wine)            0.074849            0.104749   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.015634    0.427807  9.905389               1.0  0.014056    1.672183   
1  0.015634    0.361991  9.905389               1.0  0.014056    1.510096   
2  0.012703    0.294118  8.905325               1.0  0.011276    1.369878   
3  0.012703    0.384615  8.905325               1.0  0.011276    1.554817   
4  0.010260    0.177365  1.343562               1.0  0.002624    1.055132   
5  0.010260    0.077720  1.343562               1.0  0.002624    1.021549   
6  0.011628    0.263858  1.316589               1.0  0.002796    1.086189   
7  0.011628    0.058020  1.316589               1.0  0.002796    1.014811   
8  0.010162    0.097015  1.296150               1.0  0.002322    1.024548   
9  0.010162    0.135770  1.296150               1.0  0.002322    1.035895   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.933147  0.243902   0.401979    0.394899  
1       0.939627  0.243902   0.337791    0.394899  
2       0.927778  0.200000   0.270008    0.339367  
3       0.918027  0.200000   0.356838    0.339367  
4       0.271410  0.057127   0.052252    0.127543  
5       0.294600  0.057127   0.021094    0.127543  
6       0.251547  0.049937   0.079350    0.160939  
7       0.300731  0.049937   0.014595    0.160939  
8       0.255218  0.059977   0.023960    0.116393  
9       0.246970  0.059977   0.034651    0.116393


Top ruler for the segment Alfragide


antecedents        consequents  antecedent support  \
0   (megaman zero 4)    (mashed potato)            0.014599   
1    (mashed potato)   (megaman zero 4)            0.032847   
2          (pickles)   (pokemon violet)            0.036496   
3   (pokemon violet)          (pickles)            0.032847   
4   (megaman zero 3)       (mayonnaise)            0.021898   
5       (mayonnaise)   (megaman zero 3)            0.058394   
6          (almonds)        (flax seed)            0.029197   
7        (flax seed)          (almonds)            0.047445   
8  (sparkling water)    (strong cheese)            0.076642   
9    (strong cheese)  (sparkling water)            0.018248   

   consequent support   support  confidence       lift  representativity  \
0            0.032847  0.010949    0.750000  22.833333               1.0   
1            0.014599  0.010949    0.333333  22.833333               1.0   
2            0.032847  0.010949    0.300000   9.133333               1.0   
3            0.036496  0.010949    0.333333   9.133333               1.0   
4            0.058394  0.010949    0.500000   8.562500               1.0   
5            0.021898  0.010949    0.187500   8.562500               1.0   
6            0.047445  0.010949    0.375000   7.903846               1.0   
7            0.029197  0.010949    0.230769   7.903846               1.0   
8            0.018248  0.010949    0.142857   7.828571               1.0   
9            0.076642  0.010949    0.600000   7.828571               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.010469    3.868613       0.970370  0.300000   0.741509    0.541667  
1  0.010469    1.478102       0.988679  0.300000   0.323457    0.541667  
2  0.009750    1.381648       0.924242  0.187500   0.276226    0.316667  
3  0.009750    1.445255       0.920755  0.187500   0.308081    0.316667  
4  0.009670    1.883212       0.902985  0.157895   0.468992    0.343750  
5  0.009670    1.203818       0.937984  0.157895   0.169310    0.343750  
6  0.009564    1.524088       0.899749  0.166667   0.343870    0.302885  
7  0.009564    1.262044       0.916986  0.166667   0.207634    0.302885  
8  0.009550    1.145377       0.944664  0.130435   0.126925    0.371429  
9  0.009550    2.308394       0.888476  0.130435   0.566798    0.371429

This table shows which products are frequently bought together and how strong those associations are. Includes:
- Antecedent support: Fraction of transactions containing the antecedent item(s)
- Consequent support: Fraction of transactions containing the consequent item(s)
- Support: Fraction of transactions containing both antecedent and consequent
- The confidence will give us the probability of buying the consequent given the antecedent (support/antecedent support)
- Lift: How much more likely the consequent is bought with the antecedent, compared to random
- Conviction: How strongly the antecedent implies the consequent (higher = stronger implication)  
  
And other metrics...

It is also interesting to look at the negative association, meaning two products are bought together less often than expected by chance. This could reveal products that customers see as substitutes, and preferences between segments, which is interesting for promotion planning and store layout.

In [29]:
for segment in segment_rules:
    # Get the last 10 rules (lowest lift)
    negative_rules = segment_rules[segment].tail(10)
    print(f"\nLowest lift (potential negative association) rules for the segment {segment}")
    display(negative_rules)


Lowest lift (potential negative association) rules for the segment Tech-Savvy Buyers


antecedents                    consequents  \
650                         (iMac)                     (portal 2)   
651                     (portal 2)                         (iMac)   
652                     (pancakes)                        (cream)   
653                        (cream)                     (pancakes)   
654  (gadget for tiktok streaming)                       (portal)   
655                       (portal)  (gadget for tiktok streaming)   
656                         (soda)                 (energy drink)   
657                 (energy drink)                         (soda)   
658                       (laptop)         (bluetooth headphones)   
659         (bluetooth headphones)                       (laptop)   

     antecedent support  consequent support   support  confidence      lift  \
650            0.176501            0.052637  0.010235    0.057988  1.101660   
651            0.052637            0.176501  0.010235    0.194444  1.101660   
652            0.470392            0.022977  0.011906    0.025311  1.101597   
653            0.022977            0.470392  0.011906    0.518182  1.101597   
654            0.309661            0.049922  0.017023    0.054975  1.101219   
655            0.049922            0.309661  0.017023    0.341004  1.101219   
656            0.019112            0.650235  0.013681    0.715847  1.100905   
657            0.650235            0.019112  0.013681    0.021041  1.100905   
658            0.129086            0.099948  0.014204    0.110032  1.100899   
659            0.099948            0.129086  0.014204    0.142111  1.100899   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
650               1.0  0.000944    1.005680       0.112057  0.046756   
651               1.0  0.000944    1.022274       0.097406  0.046756   
652               1.0  0.001098    1.002395       0.174141  0.024729   
653               1.0  0.001098    1.099187       0.094396  0.024729   
654               1.0  0.001565    1.005347       0.133146  0.049695   
655               1.0  0.001565    1.047563       0.096745  0.049695   
656               1.0  0.001254    1.230904       0.093442  0.020867   
657               1.0  0.001254    1.001970       0.262052  0.020867   
658               1.0  0.001302    1.011331       0.105235  0.066116   
659               1.0  0.001302    1.015182       0.101829  0.066116   

     certainty  kulczynski  
650   0.005648    0.126216  
651   0.021789    0.126216  
652   0.002389    0.271746  
653   0.090237    0.271746  
654   0.005319    0.197989  
655   0.045403    0.197989  
656   0.187589    0.368444  
657   0.001966    0.368444  
658   0.011204    0.126072  
659   0.014955    0.126072


Lowest lift (potential negative association) rules for the segment Young Promo-Sensitive


antecedents         consequents  antecedent support  \
642               (oil)          (zucchini)            0.651178   
643          (zucchini)               (oil)            0.028094   
644               (oil)      (green grapes)            0.651178   
645      (green grapes)               (oil)            0.027764   
646              (gums)       (cooking oil)            0.109656   
647       (cooking oil)              (gums)            0.463091   
648  (whole weat flour)       (cooking oil)            0.053963   
649       (cooking oil)  (whole weat flour)            0.463091   
650              (milk)       (cooking oil)            0.073076   
651       (cooking oil)              (milk)            0.463091   

     consequent support   support  confidence      lift  representativity  \
642            0.028094  0.020185    0.030997  1.103346               1.0   
643            0.651178  0.020185    0.718475  1.103346               1.0   
644            0.027764  0.019937    0.030617  1.102772               1.0   
645            0.651178  0.019937    0.718101  1.102772               1.0   
646            0.463091  0.055940    0.510143  1.101603               1.0   
647            0.109656  0.055940    0.120797  1.101603               1.0   
648            0.463091  0.027517    0.509924  1.101130               1.0   
649            0.053963  0.027517    0.059420  1.101130               1.0   
650            0.463091  0.037238    0.509583  1.100394               1.0   
651            0.073076  0.037238    0.080413  1.100394               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
642  0.001891    1.002996       0.268522  0.030625   0.002987    0.374736  
643  0.001891    1.239044       0.096374  0.030625   0.192926    0.374736  
644  0.001858    1.002943       0.267168  0.030254   0.002935    0.374359  
645  0.001858    1.237400       0.095855  0.030254   0.191854    0.374359  
646  0.005159    1.096052       0.103592  0.108242   0.087634    0.315470  
647  0.005159    1.012672       0.171784  0.108242   0.012514    0.315470  
648  0.002527    1.095562       0.097081  0.056210   0.087226    0.284672  
649  0.002527    1.005802       0.171057  0.056210   0.005769    0.284672  
650  0.003397    1.094800       0.098428  0.074637   0.086591    0.294998  
651  0.003397    1.007978       0.169926  0.074637   0.007915    0.294998


Lowest lift (potential negative association) rules for the segment Promo-Oriented Seniors


antecedents         consequents  antecedent support  \
376  (whole wheat rice)              (gums)            0.072813   
377              (gums)  (whole wheat rice)            0.275182   
378              (eggs)              (gums)            0.052583   
379              (gums)              (eggs)            0.275182   
380           (napkins)             (cream)            0.205470   
381             (cream)           (napkins)            0.079791   
382          (pet food)              (milk)            0.102629   
383              (milk)          (pet food)            0.099246   
384           (cookies)           (ketchup)            0.079862   
385           (ketchup)           (cookies)            0.136392   

     consequent support   support  confidence      lift  representativity  \
376            0.275182  0.022062    0.303001  1.101095               1.0   
377            0.072813  0.022062    0.080174  1.101095               1.0   
378            0.275182  0.015930    0.302949  1.100906               1.0   
379            0.052583  0.015930    0.057889  1.100906               1.0   
380            0.079791  0.018045    0.087822  1.100641               1.0   
381            0.205470  0.018045    0.226148  1.100641               1.0   
382            0.099246  0.011207    0.109203  1.100332               1.0   
383            0.102629  0.011207    0.112926  1.100332               1.0   
384            0.136392  0.011983    0.150044  1.100091               1.0   
385            0.079862  0.011983    0.087855  1.100091               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
376  0.002026    1.039913       0.099023  0.067690   0.038381    0.191588  
377  0.002026    1.008003       0.126670  0.067690   0.007939    0.191588  
378  0.001460    1.039836       0.096745  0.051085   0.038310    0.180419  
379  0.001460    1.005632       0.126456  0.051085   0.005600    0.180419  
380  0.001650    1.008803       0.115085  0.067528   0.008727    0.156985  
381  0.001650    1.026722       0.099367  0.067528   0.026026    0.156985  
382  0.001022    1.011178       0.101612  0.058780   0.011055    0.111065  
383  0.001022    1.011608       0.101230  0.058780   0.011475    0.111065  
384  0.001090    1.016062       0.098881  0.058661   0.015808    0.118950  
385  0.001090    1.008763       0.105354  0.058661   0.008687    0.118950


Lowest lift (potential negative association) rules for the segment Wide-Range Shoppers


antecedents             consequents  antecedent support  \
480      (whole weat flour)                   (ham)            0.046205   
481                   (ham)      (whole weat flour)            0.229439   
482        (barbecue sauce)  (extra dark chocolate)            0.498284   
483  (extra dark chocolate)        (barbecue sauce)            0.028251   
484           (cooking oil)               (cookies)            0.227261   
485               (cookies)           (cooking oil)            0.052475   
486                   (ham)         (grated cheese)            0.229439   
487         (grated cheese)                   (ham)            0.041056   
488           (cooking oil)                 (cream)            0.227261   
489                 (cream)           (cooking oil)            0.051221   

     consequent support   support  confidence      lift  representativity  \
480            0.229439  0.011683    0.252857  1.102067               1.0   
481            0.046205  0.011683    0.050921  1.102067               1.0   
482            0.028251  0.015512    0.031130  1.101913               1.0   
483            0.498284  0.015512    0.549065  1.101913               1.0   
484            0.052475  0.013135    0.057798  1.101442               1.0   
485            0.227261  0.013135    0.250314  1.101442               1.0   
486            0.041056  0.010363    0.045167  1.100125               1.0   
487            0.229439  0.010363    0.252412  1.100125               1.0   
488            0.051221  0.012805    0.056346  1.100058               1.0   
489            0.227261  0.012805    0.250000  1.100058               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
480  0.001082    1.031344       0.097101  0.044261   0.030391    0.151889  
481  0.001082    1.004969       0.120191  0.044261   0.004944    0.151889  
482  0.001435    1.002972       0.184342  0.030354   0.002963    0.290098  
483  0.001435    1.112614       0.095176  0.030354   0.101216    0.290098  
484  0.001210    1.005650       0.119185  0.049270   0.005618    0.154056  
485  0.001210    1.030751       0.097200  0.049270   0.029834    0.154056  
486  0.000943    1.004305       0.118112  0.039838   0.004287    0.148789  
487  0.000943    1.030729       0.094909  0.039838   0.029813    0.148789  
488  0.001165    1.005431       0.117707  0.048199   0.005402    0.153173  
489  0.001165    1.030319       0.095868  0.048199   0.029427    0.153173


Lowest lift (potential negative association) rules for the segment Vegetarians


antecedents          consequents  antecedent support  \
262               (mint)  (frozen vegetables)            0.051202   
263  (frozen vegetables)               (mint)            0.222994   
264       (strawberries)        (green beans)            0.107165   
265        (green beans)       (strawberries)            0.179363   
266      (mashed potato)      (mineral water)            0.349516   
267      (mineral water)      (mashed potato)            0.037543   
268         (shower gel)           (hot dogs)            0.116297   
269           (hot dogs)         (shower gel)            0.082345   
270      (herb & pepper)        (green beans)            0.093740   
271        (green beans)      (herb & pepper)            0.179363   

     consequent support   support  confidence      lift  representativity  \
262            0.222994  0.012566    0.245427  1.100598               1.0   
263            0.051202  0.012566    0.056353  1.100598               1.0   
264            0.179363  0.021152    0.197378  1.100438               1.0   
265            0.107165  0.021152    0.117929  1.100438               1.0   
266            0.037543  0.014440    0.041313  1.100423               1.0   
267            0.349516  0.014440    0.384615  1.100423               1.0   
268            0.082345  0.010537    0.090604  1.100302               1.0   
269            0.116297  0.010537    0.127962  1.100302               1.0   
270            0.179363  0.018498    0.197336  1.100202               1.0   
271            0.093740  0.018498    0.103133  1.100202               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
262  0.001149    1.029729       0.096336  0.048031   0.028871    0.150890  
263  0.001149    1.005458       0.117635  0.048031   0.005429    0.150890  
264  0.001931    1.022445       0.102226  0.079706   0.021952    0.157653  
265  0.001931    1.012203       0.111220  0.079706   0.012055    0.157653  
266  0.001318    1.003933       0.140293  0.038752   0.003917    0.212964  
267  0.001318    1.057036       0.094818  0.038752   0.053959    0.212964  
268  0.000961    1.009082       0.103155  0.056017   0.009000    0.109283  
269  0.000961    1.013377       0.099339  0.056017   0.013200    0.109283  
270  0.001685    1.022391       0.100496  0.072655   0.021901    0.150234  
271  0.001685    1.010473       0.110982  0.072655   0.010364    0.150234


Lowest lift (potential negative association) rules for the segment Outliers


antecedents                    consequents  \
862                   (fresh tuna)                     (escalope)   
863                     (escalope)                   (fresh tuna)   
864                    (champagne)                     (zucchini)   
865                     (zucchini)                    (champagne)   
866                   (fresh tuna)               (barbecue sauce)   
867               (barbecue sauce)                   (fresh tuna)   
868  (gadget for tiktok streaming)                    (champagne)   
869                    (champagne)  (gadget for tiktok streaming)   
870               (vacuum cleaner)                   (fresh tuna)   
871                   (fresh tuna)               (vacuum cleaner)   

     antecedent support  consequent support   support  confidence      lift  \
862            0.226442            0.059125  0.014987    0.066183  1.119368   
863            0.059125            0.226442  0.014987    0.253472  1.119368   
864            0.322829            0.035619  0.012831    0.039746  1.115856   
865            0.035619            0.322829  0.012831    0.360231  1.115856   
866            0.226442            0.103162  0.025970    0.114687  1.111724   
867            0.103162            0.226442  0.025970    0.251741  1.111724   
868            0.081297            0.322829  0.029049    0.357323  1.106850   
869            0.322829            0.081297  0.029049    0.089984  1.106850   
870            0.041983            0.226442  0.010470    0.249389  1.101335   
871            0.226442            0.041983  0.010470    0.046238  1.101335   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
862               1.0  0.001598    1.007558       0.137855  0.055387   
863               1.0  0.001598    1.036208       0.113340  0.055387   
864               1.0  0.001332    1.004297       0.153324  0.037125   
865               1.0  0.001332    1.058461       0.107662  0.037125   
866               1.0  0.002610    1.013019       0.129914  0.085531   
867               1.0  0.002610    1.033811       0.112056  0.085531   
868               1.0  0.002804    1.053673       0.105078  0.077449   
869               1.0  0.002804    1.009546       0.142556  0.077449   
870               1.0  0.000963    1.030570       0.096043  0.040589   
871               1.0  0.000963    1.004461       0.118945  0.040589   

     certainty  kulczynski  
862   0.007501    0.159828  
863   0.034942    0.159828  
864   0.004279    0.199988  
865   0.055232    0.199988  
866   0.012851    0.183214  
867   0.032705    0.183214  
868   0.050939    0.223654  
869   0.009455    0.223654  
870   0.029664    0.147813  
871   0.004441    0.147813


Lowest lift (potential negative association) rules for the segment Large Family Households


antecedents          consequents  antecedent support  \
280               (eggs)    (ratchet & clank)            0.094914   
281    (ratchet & clank)               (eggs)            0.568102   
282           (hot dogs)  (ratchet & clank 2)            0.039527   
283  (ratchet & clank 2)           (hot dogs)            0.256105   
284             (turkey)    (ratchet & clank)            0.108447   
285    (ratchet & clank)             (turkey)            0.568102   
286    (ratchet & clank)    (chocolate bread)            0.568102   
287    (chocolate bread)    (ratchet & clank)            0.018630   
288      (grated cheese)           (pet food)            0.045569   
289           (pet food)      (grated cheese)            0.298842   

     consequent support   support  confidence      lift  representativity  \
280            0.568102  0.059353    0.625332  1.100739               1.0   
281            0.094914  0.059353    0.104476  1.100739               1.0   
282            0.256105  0.011140    0.281847  1.100513               1.0   
283            0.039527  0.011140    0.043500  1.100513               1.0   
284            0.568102  0.067787    0.625073  1.100283               1.0   
285            0.108447  0.067787    0.119322  1.100283               1.0   
286            0.018630  0.011644    0.020496  1.100155               1.0   
287            0.568102  0.011644    0.625000  1.100155               1.0   
288            0.298842  0.014980    0.328729  1.100011               1.0   
289            0.045569  0.014980    0.050126  1.100011               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
280  0.005432    1.152748       0.101117  0.098321   0.132508    0.364904  
281  0.005432    1.010677       0.211900  0.098321   0.010564    0.364904  
282  0.001017    1.035845       0.095091  0.039159   0.034604    0.162673  
283  0.001017    1.004154       0.122777  0.039159   0.004136    0.162673  
284  0.006178    1.151952       0.102229  0.111352   0.131908    0.372197  
285  0.006178    1.012349       0.211028  0.111352   0.012198    0.372197  
286  0.001060    1.001905       0.210784  0.020247   0.001901    0.322748  
287  0.001060    1.151729       0.092766  0.020247   0.131740    0.322748  
288  0.001362    1.044524       0.095259  0.045472   0.042626    0.189428  
289  0.001362    1.004798       0.129668  0.045472   0.004775    0.189428


Lowest lift (potential negative association) rules for the segment High-Spending Shoppers


antecedents             consequents  antecedent support  \
486       (frozen smoothie)    (final fantasy XXII)            0.294704   
487    (final fantasy XXII)       (frozen smoothie)            0.032539   
488        (vacuum cleaner)               (avocado)            0.043678   
489               (avocado)        (vacuum cleaner)            0.254153   
490               (avocado)            (ring light)            0.254153   
491            (ring light)               (avocado)            0.052765   
492       (frozen smoothie)         (fromage blanc)            0.294704   
493         (fromage blanc)       (frozen smoothie)            0.146766   
494  (bluetooth headphones)            (shower gel)            0.436682   
495            (shower gel)  (bluetooth headphones)            0.072210   

     consequent support   support  confidence      lift  representativity  \
486            0.032539  0.010553    0.035809  1.100509               1.0   
487            0.294704  0.010553    0.324324  1.100509               1.0   
488            0.254153  0.012214    0.279642  1.100291               1.0   
489            0.043678  0.012214    0.048058  1.100291               1.0   
490            0.052765  0.014755    0.058055  1.100242               1.0   
491            0.254153  0.014755    0.279630  1.100242               1.0   
492            0.146766  0.047586    0.161472  1.100204               1.0   
493            0.294704  0.047586    0.324234  1.100204               1.0   
494            0.072210  0.034688    0.079436  1.100067               1.0   
495            0.436682  0.034688    0.480379  1.100067               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
486  0.000964    1.003392       0.129491  0.033323   0.003380    0.180067  
487  0.000964    1.043838       0.094401  0.033323   0.041997    0.180067  
488  0.001113    1.035384       0.095313  0.042764   0.034175    0.163850  
489  0.001113    1.004602       0.122209  0.042764   0.004581    0.163850  
490  0.001344    1.005615       0.122155  0.050502   0.005584    0.168842  
491  0.001344    1.035366       0.096184  0.050502   0.034158    0.168842  
492  0.004334    1.017538       0.129134  0.120814   0.017236    0.242853  
493  0.004334    1.043699       0.106744  0.120814   0.041870    0.242853  
494  0.003155    1.007849       0.161479  0.073151   0.007788    0.279908  
495  0.003155    1.084094       0.098044  0.073151   0.077571    0.279908


Lowest lift (potential negative association) rules for the segment Alfragide


antecedents      consequents  antecedent support  \
1252        (oatmeal)           (salt)            0.083942   
1253           (salt)        (oatmeal)            0.156934   
1254           (beer)      (meatballs)            0.149635   
1255      (meatballs)           (beer)            0.551095   
1256           (salt)  (mineral water)            0.156934   
1257  (mineral water)           (salt)            0.463504   
1258           (salt)        (seabass)            0.156934   
1259        (seabass)           (salt)            0.189781   
1260           (eggs)          (pasta)            0.175182   
1261          (pasta)           (eggs)            0.321168   

      consequent support   support  confidence      lift  representativity  \
1252            0.156934  0.014599    0.173913  1.108190               1.0   
1253            0.083942  0.014599    0.093023  1.108190               1.0   
1254            0.551095  0.091241    0.609756  1.106445               1.0   
1255            0.149635  0.091241    0.165563  1.106445               1.0   
1256            0.463504  0.080292    0.511628  1.103827               1.0   
1257            0.156934  0.080292    0.173228  1.103827               1.0   
1258            0.189781  0.032847    0.209302  1.102862               1.0   
1259            0.156934  0.032847    0.173077  1.102862               1.0   
1260            0.321168  0.062044    0.354167  1.102746               1.0   
1261            0.175182  0.062044    0.193182  1.102746               1.0   

      leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
1252  0.001425    1.020553       0.106574  0.064516   0.020139    0.133468  
1253  0.001425    1.010013       0.115801  0.064516   0.009914    0.133468  
1254  0.008778    1.150319       0.113133  0.149701   0.130676    0.387660  
1255  0.008778    1.019088       0.214309  0.149701   0.018731    0.387660  
1256  0.007552    1.098540       0.111570  0.148649   0.089701    0.342428  
1257  0.007552    1.019708       0.175325  0.148649   0.019327    0.342428  
1258  0.003064    1.024689       0.110630  0.104651   0.024094    0.191190  
1259  0.003064    1.019521       0.115115  0.104651   0.019148    0.191190  
1260  0.005781    1.051095       0.112962  0.142857   0.048611    0.273674  
1261  0.005781    1.022309       0.137255  0.142857   0.021822    0.273674

### Targeted Promotions

**Tech-Savvy Buyers**
This group shows strong associations between premium or niche alcoholic beverages such as white wine, cider, dessert wine, champagne and bramble. Their high lift values and confidence scores suggest cross-category buying behaviors among trendy or gourmet drink options. Since these customers are likely to be active on mobile apps or digital platforms we will focus on digital, experiential, and premium promotion mechanics.  
Our promotion idea for this segment is a  **“Curated Experience Box”**:  
- Includes: 1 bottle of white wine, 1 bottle of cider, gourmet sides (e.g., olives, cheese).
- The packaging includes a QR code that when scanned with the supermarket’s app gives access to a virtual tasting experience + 10% off other alcoholic beverages.
- Delivery Channel: Mobile App + Web platform (with push notifications and gamification) and limited-time offer with countdown timers.
  
**Young Promo-Sensitive**
These shoppers are highly sensitive to deals and promotions, so it is important to give them different promotions. We found positive patterns like: 
- Barbecue Sauce + Deodorant/Cologne (Lift > 5)
- Salt + Beer / White Wine  (Lift > 4)
- Cider + White Wine
And negative association between cooking staples like oil, whole wheat flour, zucchini, green grapes, indicating "Healthy image" product bundles are not appealing to this segment.  
Their behavior indicates impulse-driven buying when promotions are present and a strong tendency to respond to combo offers, especially involving categories like personal care (deodorant, cologne) and leisure items (barbecue sauce, beer, cider). Strong crossover between lifestyle and grooming suggests a social/party persona, which encourages multi-category purchases with a fun and event-oriented framing. Negative associations show low interest in cooking basics and wellness foods (e.g., oil, zucchini, whole wheat flour), indicating they may skip health staples in favor of indulgent or trendy purchases.   

Our promotion idea for this segment is a **“Grill Essentials Kit”**:
- The bundle includes 1 BBQ Sauce, salt and a pack of beers.
- Near the “Grill Essentials Kit” we will place a deodorant + cologne bundle that will be 30% off when bought together with the grill bundle.
- Delivery channel: App + in-store endcap displays

For this group, the  “Curated Experience Box” promotion will also be interesting given their high interest and association between alcoholic beverages.  

**Promo-Oriented Seniors**
Despite being older, this segment exhibits similar behaviors to the Young Promo-Sensitive group, responding well to indulgent, social, or event-focused promotions. The “Curated Experience Box”, and “Grill Essentials Kit” will be good promotions for this group since they show positive association between alcoholic beverages and grill products like BBQ sauce, salt and beer. They also show association between pantry purchases like cereals, butter, or fresh bread, and pet food.  
Our promotion idea for this segment is a **20% discount on petfood when spending on pantry essential**:
- Spend over 10 euros on pantry essentials (bread, cereal, butter…) and get 20% off on pet food.
- Delivery Channels: In-store signage near relevant sections + optional digital coupon via app (although this group isn't especially prone to app usage).
We also propose to set a sauce stand near fries, since mayo and fries seems to have a positive association for this segment.  

**Wide-Range Shoppers**
This segment likes alcohol variety, suggesting a refined or exploratory drink preference likely social, occasion-based, or hobbyist consumption. They're less attracted to essentials like flour, oil, or odd mixes like barbecue sauce with dark chocolate. The associations between white wine, cider, dessert wine, beer and bramble show a very high lift (>8). Seeing the interest this group and others have on alcoholic beverages, we suggest another alcohol based promotion.
Our promotion idea for this segment (and other alcohol indulging segments) is a **“Buy 2, Get 1 Free Happy Hour”**:
- From 17:00 to 19:00 Monday through Friday customers can get a free bottle when buying 2 other alcoholic beverages.
- Delivery Channels: In-store posters + app push notifications during active hours.
  
**Vegetarians**
This group shows health-forward eating habits, with high association between salad, flax seed, green grapes, melons.  Notably, there are also links to personal care products (e.g., body spray, shower gel, cologne), showing a self-care mindset.  
Our promotion idea for this segment is a **“Salad Builder”** bundle:
- Salad bundle: prewashed salad mix, flax seeds and cherry tomatoes.
- Fruit salad bundle: green grapes, melons and strawberries.
- Buy one of the bundles and get the other 25% off.  
- Also provide a “Save it!” discount when bundles are near the expiration date.
- Delivery Channels: Visual-focused in-store signage + health-themed app messaging.  
Framing should emphasize nourishment, convenience, and freshness.  
  
**Large Family Households**
This segment includes large families with kids, which shows in the association rules where food staples like milk, cooking oil or burgers are associated with each other and also with games for the youngest like Pokemon Violet.  
Our promotion idea for this segment is a  **“Grocery + Game Combo”**:
- Spend €50+ on groceries and receive 20% off a videogame of your choice.
- Large families show a high spend on groceries and this will promote videogame purchases like Pokemon Violet or Metroid Fusion.
- Delivery Channels: App-exclusive promotion + targeted posters in-store (e.g., near game section or kids' products).
  
**High Spending Shoppers**
These shoppers enjoy gourmet products and often purchase together products like cider/dessert wine and white whine, meatballs and blueberries, shrimp and french whine… Judging by their premium lifestyle these buyers won’t focus as much on promotions, but will enjoy previous alcoholic beverages promotions: “Curated Experience Box” and “Buy 2, Get 1 Free Happy Hour”.
  
**Alfragide**
The Alfragide segment exhibits some of the highest lift association rules among all segments, suggesting a shopper profile that is: Quirky and niche, interested in premium tech and food and willing to mix luxury with practicality. Interestingly, the key insights we saw are:
- Magman Zero 4 → Mashed Potato (Lift = 22.83)
- Pickles → Pokémon Violet (Lift = 9.13)
- Megaman Zero 3 → Mayonnaise (Lift = 8.56)
For this segment we believe our previous “Grocery + Game Combo” would be the most beneficial while the delivery method can include posters in Alfragide store.

**Outliers**
This group reflects outliers we found in the data and the rules we find for them are interesting across all segments and stores.The rules we saw highlight non-mainstream, or unexpected item pairings with strong lift. Some main associations found include:
- Pancakes → Energy Drink
- Minecraft → Ratchet & Clank
- Babies Food → Ratchet & Clank
- Minecraft → Babies Food
Again, we can see a pattern of videogame purchases. These people would also benefit from our  “Grocery + Game Combo” previously mentioned.
